pkg/client/cloudevents/grpcsource/pager.go

```diff
@@ -0,0 +1,107 @@
+package grpcsource
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/openshift-online/maestro/pkg/api/openapi"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+)
+
+// MaxListPageSize is the maximum size of one page, default is 400.
+// NOTE: This should be reset carefully, when increasing this value, both maestro server's memory limit
+// and the page size of resources need to be considered, if a bigger value is used, it might lead to
+// maestro server OOM.
+var MaxListPageSize int32 = 400
+
+// pageList assists client code in breaking large list queries into multiple smaller chunks of PageSize or smaller.
+func pageList(client *openapi.APIClient, search string, opts metav1.ListOptions) (*openapi.ResourceBundleList, string, error) {
+	items := []openapi.ResourceBundle{}
+
+	page, err := page(opts)
+	if err != nil {
+		return nil, "", err
+	}
+
+	limit := opts.Limit
+	if limit < 0 {
+		return nil, "", fmt.Errorf("limit cannot be less than 0")
+	}
+
+	var total int32 = 0
+	nextPage := ""
+	pageSize := pageSize(int32(limit))
+	offset := (page - 1) * pageSize
+	for {
+		klog.V(4).Infof("list works with search=%s, page=%d, size=%d", search, page, page)
+		rbs, _, err := client.DefaultApi.ApiMaestroV1ResourceBundlesGet(context.Background()).
+			Search(search).
+			Page(page).
+			Size(pageSize).
+			Execute()
+		if err != nil {
+			return nil, "", err
+		}
+		klog.V(4).Infof("listed works total=%d, size=%d, page=%d", rbs.Total, rbs.Size, rbs.Page)
+
+		items = append(items, rbs.Items...)
+		total = rbs.Size + total
+		page = page + 1
+
+		if rbs.Size < pageSize {
+			// reaches the last page, stop list
+			if (total + offset) < rbs.Total {
```

Comment:
why do we need this if? this is the last page, shouldn't we just break?

Reply:
yes, this judgment is needless